File: php/CreoleCentricAPI.php
<?php
/**
 * CreoleCentric TTS API Client Example
 * =====================================
 * This example demonstrates how to use the CreoleCentric Text-to-Speech API.
 *
 * Requirements:
 *     composer require guzzlehttp/guzzle vlucas/phpdotenv
 */

require_once __DIR__ . '/vendor/autoload.php';

use GuzzleHttp\Client;
use GuzzleHttp\Exception\GuzzleException;

/**
 * Client for interacting with CreoleCentric TTS API
 */
class CreoleCentricAPI
{
    private string $apiKey;
    private string $baseUrl;
    private Client $client;
    private array $headers;

    /**
     * Initialize the API client
     *
     * @param string $apiKey Your API key starting with 'cc_'
     * @param string $baseUrl Base URL for the API (default: production URL)
     */
    public function __construct(string $apiKey, string $baseUrl = 'https://creolecentric.com/api/v1')
    {
        $this->apiKey = $apiKey;
        $this->baseUrl = rtrim($baseUrl, '/');
        $this->headers = [
            'Authorization' => "ApiKey {$apiKey}",
            'Content-Type' => 'application/json',
            'Accept' => 'application/json'
        ];

        $this->client = new Client([
            'base_uri' => $this->baseUrl,
            'timeout' => 30.0,
            'headers' => $this->headers
        ]);
    }

    /**
     * Make an HTTP request to the API
     *
     * @param string $method HTTP method (GET, POST, etc.)
     * @param string $endpoint API endpoint (e.g., '/tts/jobs/')
     * @param array|null $data JSON data for POST requests
     * @param array|null $params Query parameters
     * @return array Response JSON data
     * @throws GuzzleException If the request fails
     */
    private function makeRequest(string $method, string $endpoint, ?array $data = null, ?array $params = null): array
    {
        try {
            $options = [];

            if ($data !== null) {
                $options['json'] = $data;
            }

            if ($params !== null) {
                $options['query'] = $params;
            }

            $response = $this->client->request($method, $endpoint, $options);
            $body = (string) $response->getBody();

            return $body ? json_decode($body, true) : [];
        } catch (GuzzleException $e) {
            echo "HTTP Error: " . $e->getMessage() . "\n";
            if ($e->hasResponse()) {
                echo "Response: " . $e->getResponse()->getBody() . "\n";
            }
            throw $e;
        }
    }

    // ============== Health Check ==============

    /**
     * Check API health status
     *
     * @return array Health status information
     */
    public function checkHealth(): array
    {
        return $this->makeRequest('GET', '/health/');
    }

    // ============== User & Credits ==============

    /**
     * Get current user information
     *
     * @return array User profile data
     */
    public function getUserInfo(): array
    {
        return $this->makeRequest('GET', '/users/profile/');
    }

    /**
     * Get current credit balance
     *
     * @return array Credit balance information
     */
    public function getCreditBalance(): array
    {
        return $this->makeRequest('GET', '/credits/balance/');
    }

    // ============== Voices & Models ==============

    /**
     * Get list of available voices
     *
     * @return array Dictionary containing:
     *   - success: bool
     *   - voices: array of voice dictionaries
     *   - count: number of voices
     *   - source: data source (infer or local)
     */
    public function getVoices(): array
    {
        return $this->makeRequest('GET', '/tts/voices/');
    }

    /**
     * Get list of available TTS models
     *
     * @return array Dictionary containing:
     *   - success: bool
     *   - models: array of model dictionaries
     *   - count: number of models
     */
    public function getModels(): array
    {
        return $this->makeRequest('GET', '/tts/models/');
    }

    /**
     * Get voice settings configuration
     *
     * @return array Voice settings data
     */
    public function getVoiceSettings(): array
    {
        return $this->makeRequest('GET', '/tts/voice-settings/');
    }

    // ============== TTS Jobs ==============

    /**
     * Create a new TTS job
     *
     * @param string $text Text to convert to speech
     * @param string $voiceId ID of the voice to use
     * @param string $modelId ID of the model to use
     * @param array $options Additional parameters (speed, pitch, webhook_url, etc.)
     * @return array Job details including job_id
     */
    public function createTtsJob(
        string $text,
        string $voiceId = 'qW6MAd7f5iuYw7bAH96wC',
        string $modelId = 'ccl_ht_v100',
        array $options = []
    ): array {
        $data = array_merge([
            'text' => $text,
            'voice_id' => $voiceId,
            'model_id' => $modelId
        ], $options);

        return $this->makeRequest('POST', '/tts/jobs/', $data);
    }

    /**
     * Get status of a TTS job
     *
     * @param string $jobId UUID of the job
     * @return array Job status and details
     */
    public function getJobStatus(string $jobId): array
    {
        return $this->makeRequest('GET', "/tts/jobs/{$jobId}/status/");
    }

    /**
     * Get full details of a TTS job
     *
     * @param string $jobId UUID of the job
     * @return array Complete job information
     */
    public function getJobDetails(string $jobId): array
    {
        return $this->makeRequest('GET', "/tts/jobs/{$jobId}/");
    }

    /**
     * List TTS jobs for the current user
     *
     * @param int $limit Number of jobs to return
     * @param int $offset Pagination offset
     * @return array List of jobs with pagination info
     */
    public function listJobs(int $limit = 10, int $offset = 0): array
    {
        return $this->makeRequest('GET', '/tts/jobs/list/', null, [
            'limit' => $limit,
            'offset' => $offset
        ]);
    }

    /**
     * Cancel a pending or processing TTS job
     *
     * @param string $jobId UUID of the job to cancel
     * @return array Cancellation status
     */
    public function cancelJob(string $jobId): array
    {
        return $this->makeRequest('POST', "/tts/jobs/{$jobId}/cancel/");
    }

    // ============== Express TTS ==============

    /**
     * Use express TTS for immediate audio generation (shorter texts)
     *
     * @param string $text Text to convert (max 500 characters recommended)
     * @param string $voiceId Voice to use
     * @return string Audio data as binary string
     * @throws GuzzleException
     */
    public function expressTts(string $text, string $voiceId = 'qW6MAd7f5iuYw7bAH96wC'): string
    {
        $data = [
            'text' => $text,
            'voice_id' => $voiceId
        ];

        $response = $this->client->post('/tts/express/', [
            'json' => $data
        ]);

        return (string) $response->getBody();
    }
}

/**
 * Example usage of the CreoleCentric API client
 */
function main(): void
{
    // Load environment variables from .env file
    if (file_exists(__DIR__ . '/.env')) {
        $dotenv = Dotenv\Dotenv::createImmutable(__DIR__);
        $dotenv->load();
    }

    // Load API key from environment variable
    $apiKey = $_ENV['CREOLECENTRIC_API_KEY'] ?? getenv('CREOLECENTRIC_API_KEY');

    if (!$apiKey) {
        echo "Error: CREOLECENTRIC_API_KEY environment variable not set\n";
        echo "Please set it with your API key: export CREOLECENTRIC_API_KEY='cc_your_key_here'\n";
        return;
    }

    try {
        // Initialize client
        $client = new CreoleCentricAPI($apiKey);

        // 1. Check API health
        echo str_repeat('=', 50) . "\n";
        echo "1. Checking API Health\n";
        echo str_repeat('=', 50) . "\n";
        $health = $client->checkHealth();
        echo "API Status: " . ($health['status'] ?? 'unknown') . "\n";
        echo "Version: " . ($health['version'] ?? 'unknown') . "\n";
        echo "\n";

        // 2. Get credit balance
        echo str_repeat('=', 50) . "\n";
        echo "2. Credit Balance\n";
        echo str_repeat('=', 50) . "\n";
        $balance = $client->getCreditBalance();
        echo "Total Credits: " . number_format($balance['total_credits'] ?? 0) . "\n";
        echo "Subscription Credits: " . number_format($balance['subscription_credits'] ?? 0) . "\n";
        echo "Purchased Credits: " . number_format($balance['purchased_credits'] ?? 0) . "\n";
        echo "\n";

        // 3. Get available voices
        echo str_repeat('=', 50) . "\n";
        echo "3. Available Voices\n";
        echo str_repeat('=', 50) . "\n";
        $voicesResponse = $client->getVoices();
        $voices = $voicesResponse['voices'] ?? [];
        $count = $voicesResponse['count'] ?? count($voices);
        $source = $voicesResponse['source'] ?? 'unknown';
        echo "Found {$count} voices (source: {$source}):\n";
        foreach (array_slice($voices, 0, 5) as $voice) {
            echo "  - " . ($voice['name'] ?? 'Unknown') . " (ID: " . ($voice['voice_id'] ?? 'N/A') . ")\n";
            echo "    Region: " . ($voice['region'] ?? 'N/A') . ", Gender: " . ($voice['gender'] ?? 'N/A') . "\n";
        }
        echo "\n";

        // 4. Get available models
        echo str_repeat('=', 50) . "\n";
        echo "4. Available Models\n";
        echo str_repeat('=', 50) . "\n";
        $modelsResponse = $client->getModels();
        $models = $modelsResponse['models'] ?? [];
        $count = $modelsResponse['count'] ?? count($models);
        echo "Found {$count} models:\n";
        foreach ($models as $model) {
            $name = $model['display_name'] ?? $model['name'] ?? 'Unknown';
            echo "  - {$name} (ID: " . ($model['id'] ?? 'N/A') . ")\n";
            echo "    Description: " . ($model['description'] ?? 'N/A') . "\n";
        }
        echo "\n";

        // 5. Create a TTS job
        echo str_repeat('=', 50) . "\n";
        echo "5. Creating TTS Job\n";
        echo str_repeat('=', 50) . "\n";

        $text = 'Bonjou! Mwen se yon egzanp API pou CreoleCentric. Mwen ka pale Kreyòl ayisyen.';
        echo "Text: {$text}\n";

        // Use Xavier Bruneau voice and default Haitian Creole model
        // To find voice IDs: Go to Voice Library page, click "More" (...) on any voice card
        // To find model IDs: In TTS interface, go to Speech Options tab, click Model field

        // Use a real voice ID - don't rely on voices list which may contain placeholders
        $voiceId = 'i4mRPwKM2yHwXhbmkN514';  // Xavier Bruneau
        $modelId = 'ccl_ht_v100';  // Default Haitian Creole model

        // If you want to use a voice from the list, make sure it's not a placeholder
        if (!empty($voices) && !in_array($voices[0]['voice_id'] ?? '', ['voice_1', 'voice_2'])) {
            $voiceId = $voices[0]['voice_id'];
        }
        if (!empty($models)) {
            $modelId = $models[0]['id'];
        }

        // To use webhooks, add webhook_url parameter:
        $job = $client->createTtsJob(
            $text,
            $voiceId,
            $modelId,
            ['webhook_url' => 'https://your-app.com/webhooks/tts']  // Your webhook endpoint
        );

        $jobId = $job['id'] ?? 'N/A';
        echo "Job created successfully!\n";
        echo "Job ID: {$jobId}\n";
        echo "Status: " . ($job['status'] ?? 'unknown') . "\n";
        echo "Credits used: " . ($job['credits_used'] ?? 0) . "\n";
        echo "\n";
        echo "📢 Webhook notifications will be sent to your endpoint:\n";
        echo "   - tts_queued → tts_started → tts_synthesized → tts_uploaded → tts_delivered\n";
        echo "   See examples/webhook_server.php for webhook handling example\n";
        echo "\n";

        // 6. List recent jobs
        echo str_repeat('=', 50) . "\n";
        echo "6. Recent Jobs\n";
        echo str_repeat('=', 50) . "\n";

        $jobs = $client->listJobs(5);
        $results = $jobs['results'] ?? [];
        echo "Recent " . count($results) . " jobs:\n";

        foreach ($results as $job) {
            $created = $job['created_at'] ?? '';
            $jobId = $job['id'] ?? '';
            $jobIdDisplay = $jobId ? substr($jobId, 0, 8) . '...' : 'N/A';
            echo "  - Job {$jobIdDisplay}\n";
            echo "    Created: {$created}\n";
            echo "    Status: " . ($job['status'] ?? 'unknown') . "\n";
            $textPreview = $job['text'] ?? '';
            if ($textPreview) {
                $preview = strlen($textPreview) > 50 ? substr($textPreview, 0, 50) . '...' : $textPreview;
                echo "    Text: {$preview}\n";
            }
        }
    } catch (Exception $e) {
        echo "Error: " . $e->getMessage() . "\n";
        echo $e->getTraceAsString() . "\n";
    }
}

// Run main function if this file is executed directly
if (php_sapi_name() === 'cli' && basename(__FILE__) === basename($_SERVER['SCRIPT_FILENAME'])) {
    main();
}

File: ruby/creolecentric_api.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

#
# CreoleCentric TTS API Client Example
# =====================================
# This example demonstrates how to use the CreoleCentric Text-to-Speech API.
#
# Requirements:
#     gem install http dotenv
#

require 'http'
require 'json'
require 'dotenv/load'

# Client for interacting with CreoleCentric TTS API
class CreoleCentricAPI
  attr_reader :api_key, :base_url

  # Initialize the API client
  #
  # @param api_key [String] Your API key starting with 'cc_'
  # @param base_url [String] Base URL for the API (default: production URL)
  def initialize(api_key, base_url = 'https://creolecentric.com/api/v1')
    @api_key = api_key
    @base_url = base_url.chomp('/')
    @headers = {
      'Authorization' => "ApiKey #{api_key}",
      'Content-Type' => 'application/json'
    }
  end

  # Make an HTTP request to the API
  #
  # @param method [Symbol] HTTP method (:get, :post, etc.)
  # @param endpoint [String] API endpoint (e.g., '/tts/jobs/')
  # @param data [Hash, nil] JSON data for POST requests
  # @param params [Hash, nil] Query parameters
  # @return [Hash] Response JSON data
  # @raise [HTTP::Error] If the request fails
  def make_request(method, endpoint, data: nil, params: nil)
    url = "#{@base_url}#{endpoint}"

    response = HTTP.headers(@headers).timeout(30).request(
      method,
      url,
      json: data,
      params: params
    )

    raise HTTP::Error, "HTTP #{response.code}: #{response.body}" unless response.status.success?

    response.body.empty? ? {} : JSON.parse(response.body.to_s)
  rescue HTTP::Error => e
    puts "HTTP Error: #{e}"
    raise
  rescue StandardError => e
    puts "Request Error: #{e}"
    raise
  end

  # ============== Health Check ==============

  # Check API health status
  #
  # @return [Hash] Health status information
  def check_health
    make_request(:get, '/health/')
  end

  # ============== User & Credits ==============

  # Get current user information
  #
  # @return [Hash] User profile data
  def get_user_info
    make_request(:get, '/users/profile/')
  end

  # Get current credit balance
  #
  # @return [Hash] Credit balance information
  def get_credit_balance
    make_request(:get, '/credits/balance/')
  end

  # ============== Voices & Models ==============

  # Get list of available voices
  #
  # @return [Hash] Dictionary containing:
  #   - success: bool
  #   - voices: list of voice dictionaries
  #   - count: number of voices
  #   - source: data source (infer or local)
  def get_voices
    make_request(:get, '/tts/voices/')
  end

  # Get list of available TTS models
  #
  # @return [Hash] Dictionary containing:
  #   - success: bool
  #   - models: list of model dictionaries
  #   - count: number of models
  def get_models
    make_request(:get, '/tts/models/')
  end

  # Get voice settings configuration
  #
  # @return [Hash] Voice settings data
  def get_voice_settings
    make_request(:get, '/tts/voice-settings/')
  end

  # ============== TTS Jobs ==============

  # Create a new TTS job
  #
  # @param text [String] Text to convert to speech
  # @param voice_id [String] ID of the voice to use
  # @param model_id [String] ID of the model to use
  # @param options [Hash] Additional parameters (speed, pitch, etc.)
  # @return [Hash] Job details including job_id
  def create_tts_job(text, voice_id: 'qW6MAd7f5iuYw7bAH96wC', model_id: 'ccl_ht_v100', **options)
    data = {
      text: text,
      voice_id: voice_id,
      model_id: model_id
    }.merge(options)

    make_request(:post, '/tts/jobs/', data: data)
  end

  # Get status of a TTS job
  #
  # @param job_id [String] UUID of the job
  # @return [Hash] Job status and details
  def get_job_status(job_id)
    make_request(:get, "/tts/jobs/#{job_id}/status/")
  end

  # Get full details of a TTS job
  #
  # @param job_id [String] UUID of the job
  # @return [Hash] Complete job information
  def get_job_details(job_id)
    make_request(:get, "/tts/jobs/#{job_id}/")
  end

  # List TTS jobs for the current user
  #
  # @param limit [Integer] Number of jobs to return
  # @param offset [Integer] Pagination offset
  # @return [Hash] List of jobs with pagination info
  def list_jobs(limit: 10, offset: 0)
    make_request(:get, '/tts/jobs/list/', params: { limit: limit, offset: offset })
  end

  # Cancel a pending or processing TTS job
  #
  # @param job_id [String] UUID of the job to cancel
  # @return [Hash] Cancellation status
  def cancel_job(job_id)
    make_request(:post, "/tts/jobs/#{job_id}/cancel/")
  end

  # ============== Express TTS ==============

  # Use express TTS for immediate audio generation (shorter texts)
  #
  # @param text [String] Text to convert (max 500 characters recommended)
  # @param voice_id [String] Voice to use
  # @return [String] Audio data as binary string
  def express_tts(text, voice_id: 'qW6MAd7f5iuYw7bAH96wC')
    data = {
      text: text,
      voice_id: voice_id
    }

    response = HTTP.headers(@headers).timeout(30).post(
      "#{@base_url}/tts/express/",
      json: data
    )

    raise HTTP::Error, "HTTP #{response.code}" unless response.status.success?

    response.body.to_s
  end
end

# Example usage of the CreoleCentric API client
def main
  # Load API key from environment variable or .env file
  api_key = ENV['CREOLECENTRIC_API_KEY']

  unless api_key
    puts 'Error: CREOLECENTRIC_API_KEY environment variable not set'
    puts "Please set it with your API key: export CREOLECENTRIC_API_KEY='cc_your_key_here'"
    return
  end

  # Initialize client
  client = CreoleCentricAPI.new(api_key)

  # 1. Check API health
  puts '=' * 50
  puts '1. Checking API Health'
  puts '=' * 50
  health = client.check_health
  puts "API Status: #{health['status']}"
  puts "Version: #{health['version']}"
  puts

  # 2. Get credit balance
  puts '=' * 50
  puts '2. Credit Balance'
  puts '=' * 50
  balance = client.get_credit_balance
  puts "Total Credits: #{balance['total_credits'] || 0}"
  puts "Subscription Credits: #{balance['subscription_credits'] || 0}"
  puts "Purchased Credits: #{balance['purchased_credits'] || 0}"
  puts

  # 3. Get available voices
  puts '=' * 50
  puts '3. Available Voices'
  puts '=' * 50
  voices_response = client.get_voices
  voices = voices_response['voices'] || []
  count = voices_response['count'] || voices.length
  source = voices_response['source'] || 'unknown'
  puts "Found #{count} voices (source: #{source}):"
  voices.first(5).each do |voice|
    puts "  - #{voice['name']} (ID: #{voice['voice_id']})"
    puts "    Region: #{voice['region']}, Gender: #{voice['gender']}"
  end
  puts

  # 4. Get available models
  puts '=' * 50
  puts '4. Available Models'
  puts '=' * 50
  models_response = client.get_models
  models = models_response['models'] || []
  count = models_response['count'] || models.length
  puts "Found #{count} models:"
  models.each do |model|
    name = model['display_name'] || model['name']
    puts "  - #{name} (ID: #{model['id']})"
    puts "    Description: #{model['description']}"
  end
  puts

  # 5. Create a TTS job
  puts '=' * 50
  puts '5. Creating TTS Job'
  puts '=' * 50

  text = 'Bonjou! Mwen se yon egzanp API pou CreoleCentric. Mwen ka pale Kreyòl ayisyen.'
  puts "Text: #{text}"

  # Use Xavier Bruneau voice and default Haitian Creole model
  # To find voice IDs: Go to Voice Library page, click "More" (...) on any voice card
  # To find model IDs: In TTS interface, go to Speech Options tab, click Model field

  # Use a real voice ID - don't rely on voices list which may contain placeholders
  voice_id = 'i4mRPwKM2yHwXhbmkN514'  # Xavier Bruneau
  model_id = 'ccl_ht_v100'  # Default Haitian Creole model

  # If you want to use a voice from the list, make sure it's not a placeholder
  if voices.any? && !%w[voice_1 voice_2].include?(voices[0]['voice_id'])
    voice_id = voices[0]['voice_id']
  end
  model_id = models[0]['id'] if models.any?

  # To use webhooks, add webhook_url parameter:
  job = client.create_tts_job(
    text,
    voice_id: voice_id,
    model_id: model_id,
    webhook_url: 'https://your-app.com/webhooks/tts'  # Your webhook endpoint
  )

  job_id = job['id']
  puts 'Job created successfully!'
  puts "Job ID: #{job_id}"
  puts "Status: #{job['status']}"
  puts "Credits used: #{job['credits_used'] || 0}"
  puts
  puts '📢 Webhook notifications will be sent to your endpoint:'
  puts '   - tts_queued → tts_started → tts_synthesized → tts_uploaded → tts_delivered'
  puts '   See examples/webhook_server.rb for webhook handling example'
  puts

  # 6. List recent jobs
  puts '=' * 50
  puts '6. Recent Jobs'
  puts '=' * 50

  jobs = client.list_jobs(limit: 5)
  results = jobs['results'] || []
  puts "Recent #{results.length} jobs:"

  results.each do |job|
    created = job['created_at'] || ''
    job_id = job['id']
    job_id_display = job_id ? "#{job_id[0..7]}..." : 'N/A'
    puts "  - Job #{job_id_display}"
    puts "    Created: #{created}"
    puts "    Status: #{job['status']}"
    text_preview = job['text'] || ''
    if !text_preview.empty?
      preview = text_preview.length > 50 ? "#{text_preview[0..49]}..." : text_preview
      puts "    Text: #{preview}"
    end
  end
rescue StandardError => e
  puts "Error: #{e}"
  puts e.backtrace
end

main if __FILE__ == $PROGRAM_NAME
